File: C++/GLA.cpp
/* Author Name: Laxmi Chari
Roll No: 22
Title: Program to implement OO Concepts using C++
Start Date: 23/07/2024,
Modified Date: 25/07/2024
Description: The program helps in understanding the structure of the Goa Legislative assembly using Object Oriented Concepts 
 */
#include <iostream>
#include <vector>
#include <string>

using namespace std;

// Abstract base class for all entities (Bills, MLAs, Officials, Sessions)
class Entity {
public:
    // Pure virtual function to be overridden by derived classes for string representation
    virtual string toString() const = 0;
    virtual ~Entity() = default;  // Virtual destructor for proper cleanup of derived classes
};

// Bill class, derived from Entity, represents a legislative bill
class Bill : public Entity {
private:
    string title;        // Title of the bill
    string description;  // Description of the bill
    bool isPassed;       // Status of the bill (passed or not)

public:
    // Constructor to initialize the title and description of the bill, sets isPassed to false
    Bill(const string &title, const string &description) : title(title), description(description), isPassed(false) {}

    // Method to pass the bill
    void pass() {
        isPassed = true;
    }

    // Override the toString method to provide a string representation of the bill
    string toString() const override {
        return "Title: " + title + "\nDescription: " + description + "\nStatus: " + (isPassed ? "Passed" : "Not Passed");
    }
};

// MLA class, derived from Entity, represents a Member of Legislative Assembly
class MLA : public Entity {
private:
    string name;          // Name of the MLA
    string constituency;  // Constituency of the MLA
    string party;         // Political party of the MLA

public:
    // Constructor to initialize the name, constituency, and party of the MLA
    MLA(const string &name, const string &constituency, const string &party) : name(name), constituency(constituency), party(party) {}

    // Override the toString method to provide a string representation of the MLA
    string toString() const override {
        return "Name: " + name + "\nConstituency: " + constituency + "\nPolitical party: " + party;
    }
};

// Official class, derived from Entity, represents an official in the assembly
class Official : public Entity {
private:
    string position;  // Position of the official
    string name;      // Name of the official

public:
    // Constructor to initialize the position and name of the official
    Official(const string &position, const string &name) : position(position), name(name) {}

    // Override the toString method to provide a string representation of the official
    string toString() const override {
        return "Position: " + position + "\nName: " + name;
    }
};

// Session class, derived from Entity, represents a session of the legislative assembly
class Session : public Entity {
private:
    string date;    // Date of the session
    string agenda;  // Agenda of the session

public:
    // Constructor to initialize the date and agenda of the session
    Session(const string &date, const string &agenda) : date(date), agenda(agenda) {}

    // Override the toString method to provide a string representation of the session
    string toString() const override {
        return "Date of the session: " + date + "\nAgenda: " + agenda;
    }
};

// GoaLegislativeAssembly class handles the management of sessions, bills, officials, and MLAs
class GoaLegislativeAssembly {
private:
    vector<Session> sessions;   // List of sessions
    vector<Bill> bills;         // List of bills
    vector<Official> officials; // List of officials
    vector<MLA> mlas;           // List of MLAs

public:
    // Menu method to display options and interact with the user
    void menu() {
        int choice;
        do {
            cout << "\nGoa Legislative Assembly Menu\n";
            cout << "1. Enter the Session Details\n";
            cout << "2. Display Session Details\n";
            cout << "3. Introduce a Bill\n";
            cout << "4. Pass a Bill\n";
            cout << "5. Display Bills\n";
            cout << "6. Add an Official\n";
            cout << "7. View Officials\n";
            cout << "8. Add an MLA\n";
            cout << "9. View MLAs\n";
            cout << "10. Exit\n";
            cout << "Enter your choice: ";
            cin >> choice;
            cin.ignore(); // Ignore the newline character left in the buffer

            // Call the appropriate method based on user's choice
            switch (choice) {
                case 1:
                    sessionDetails();
                    break;
                case 2:
                    displaySession();
                    break;
                case 3:
                    introduceBill();
                    break;
                case 4:
                    passBill();
                    break;
                case 5:
                    displayBills();
                    break;
                case 6:
                    addOfficial();
                    break;
                case 7:
                    viewOfficials();
                    break;
                case 8:
                    addMLA();
                    break;
                case 9:
                    viewMLAs();
                    break;
                case 10:
                    cout << "Exiting...\n";
                    break;
                default:
                    cout << "Invalid choice. Please try again.\n";
            }
        } while (choice != 10);  // Continue until the user chooses to exit
    }

    // Method to add session details
    void sessionDetails() {
        string date, agenda;
        cout << "Enter the Date of the session: ";
        getline(cin, date);
        cout << "Enter the Agenda of the session: ";
        getline(cin, agenda);
        sessions.emplace_back(date, agenda);  // Add the session to the list
        cout << "Session may proceed.\n";
    }

    // Method to display details of all sessions
    void displaySession() {
        if (sessions.empty()) {
            cout << "No session conducted yet.\n";
            return;
        }

        for (const auto &session : sessions) {
            cout << session.toString() << "\n\n";
        }

        // Display bills, officials, and MLAs for the session
        if (bills.empty()) {
            cout << "No bills introduced in the session.\n";
        } else {
            cout << "Bills introduced in the session:\n";
            for (const auto &bill : bills) {
                cout << bill.toString() << "\n\n";
            }
        }

        if (officials.empty()) {
            cout << "No officials to display.\n";
        } else {
            cout << "Officials present in today's Session:\n";
            for (const auto &official : officials) {
                cout << official.toString() << "\n\n";
            }
        }

        if (mlas.empty()) {
            cout << "No MLAs to display.\n";
        } else {
            cout << "List of MLAs present for today's session:\n";
            for (const auto &mla : mlas) {
                cout << mla.toString() << "\n\n";
            }
        }
    }

    // Method to introduce a new bill
    void introduceBill() {
        string title, description;
        cout << "Enter the title of the bill: ";
        getline(cin, title);
        cout << "Enter the description of the bill: ";
        getline(cin, description);
        bills.emplace_back(title, description);  // Add the bill to the list
        cout << "Bill introduced successfully.\n";
    }

    // Method to pass a bill
    void passBill() {
        if (bills.empty()) {
            cout << "No bills available to pass.\n";
            return;
        }

        string title;
        cout << "Enter the title of the bill to pass: ";
        getline(cin, title);
        // Search for the bill by title and mark it as passed
        for (auto &bill : bills) {
            if (bill.toString().find(title) != string::npos) {
                bill.pass();
                cout << "Bill passed successfully.\n";
                return;
            }
        }
        cout << "Bill not found.\n";
    }

    // Method to display all bills
    void displayBills() {
        if (bills.empty()) {
            cout << "No bills to display.\n";
            return;
        }

        for (const auto &bill : bills) {
            cout << bill.toString() << "\n\n";
        }
    }

    // Method to add an official
    void addOfficial() {
        string position, name;
        cout << "Enter the position of the official: ";
        getline(cin, position);
        cout << "Enter the name of the official: ";
        getline(cin, name);
        officials.emplace_back(position, name);  // Add the official to the list
        cout << "Official added successfully.\n";
    }

    // Method to view all officials
    void viewOfficials() {
        if (officials.empty()) {
            cout << "No officials to display.\n";
            return;
        }

        cout << "List of Officials:\n";
        for (const auto &official : officials) {
            cout << official.toString() << "\n\n";
        }
    }

    // Method to add an MLA
    void addMLA() {
        string name, constituency, party;
        cout << "Enter the name of the MLA: ";
        getline(cin, name);
        cout << "Enter the constituency of the MLA: ";
        getline(cin, constituency);
        cout << "Enter the Political party of the MLA: ";
        getline(cin, party);
        mlas.emplace_back(name, constituency, party);  // Add the MLA to the list
        cout << "MLA added successfully.\n";
    }

    // Method to view all MLAs
    void viewMLAs() {
        if (mlas.empty()) {
            cout << "No MLAs to display.\n";
            return;
        }

        cout << "List of MLAs:\n";
        for (const auto &mla : mlas) {
            cout << mla.toString() << "\n\n";
        }
    }
};

int main() {
    // Create an instance of the assembly and call the menu method to start the program
    GoaLegislativeAssembly assembly;
    assembly.menu();
    return 0;
}

File: C#/GLA.cs

/* Author Name: Laxmi Chari
Roll No: 22
Title: Program to implement OO Concepts using C#
Start Date: 22/07/2024,
Modified Date: 25/07/2024
Description: The program helps in understanding the structure of the Goa Legislative assembly using Object Oriented Concepts 
 */
using System;
using System.Collections.Generic;

// Abstract base class 'Entity' that defines the blueprint for other classes. 
// It requires every subclass to implement the ToString() method.
abstract class Entity
{
    public abstract override string ToString();
}

// The 'Bill' class represents a legislative bill.
class Bill : Entity
{
    private string title;  // Title of the bill
    private string description;  // Description of the bill
    private bool isPassed;  // Status of whether the bill is passed

    // Constructor to initialize a bill's title and description
    public Bill(string title, string description)
    {
        this.title = title;
        this.description = description;
        this.isPassed = false;  // Bill is not passed initially
    }

    // Method to mark the bill as passed
    public void Pass()
    {
        this.isPassed = true;
    }

    // Overriding the ToString() method to display the bill details
    public override string ToString()
    {
        return $"Title: {title}\nDescription: {description}\nStatus: {(isPassed ? "Passed" : "Not Passed")}";
    }
}

// The 'MLA' class represents a Member of the Legislative Assembly.
class MLA : Entity
{
    private string name;  // MLA's name
    private string constituency;  // MLA's constituency
    private string party;  // MLA's political party

    // Constructor to initialize MLA details
    public MLA(string name, string constituency, string party)
    {
        this.name = name;
        this.constituency = constituency;
        this.party = party;
    }

    // Overriding the ToString() method to display MLA details
    public override string ToString()
    {
        return $"Name: {name}\nConstituency: {constituency}\nPolitical party: {party}";
    }
}

// The 'Official' class represents an official present in the assembly.
class Official : Entity
{
    private string position;  // Position of the official
    private string name;  // Name of the official

    // Constructor to initialize the official's details
    public Official(string position, string name)
    {
        this.position = position;
        this.name = name;
    }

    // Overriding the ToString() method to display official details
    public override string ToString()
    {
        return $"Position: {position}\nName: {name}";
    }
}

// The 'Session' class represents a legislative session.
class Session : Entity
{
    private string date;  // Date of the session
    private string agenda;  // Agenda of the session

    // Constructor to initialize session details
    public Session(string date, string agenda)
    {
        this.date = date;
        this.agenda = agenda;
    }

    // Overriding the ToString() method to display session details
    public override string ToString()
    {
        return $"Date of the session: {date}\nAgenda: {agenda}";
    }
}

// Main class representing the Goa Legislative Assembly with various operations
class GoaLegislativeAssembly
{
    // Static lists to store sessions, bills, officials, and MLAs
    private static List<Session> sessions = new List<Session>();
    private static List<Bill> bills = new List<Bill>();
    private static List<Official> officials = new List<Official>();
    private static List<MLA> mlas = new List<MLA>();
    private static Scanner scanner = new Scanner();

    // Main method, entry point for the application
    public static void Main(string[] args)
    {
        int choice;
        do
        {
            // Displaying the menu options for the user
            Console.WriteLine("\nGoa Legislative Assembly Menu");
            Console.WriteLine("1. Enter the Session Details");
            Console.WriteLine("2. Display Session Details");
            Console.WriteLine("3. Introduce a Bill");
            Console.WriteLine("4. Pass a Bill");
            Console.WriteLine("5. Display Bills");
            Console.WriteLine("6. Add an Official");
            Console.WriteLine("7. View Officials");
            Console.WriteLine("8. Add an MLA");
            Console.WriteLine("9. View MLAs");
            Console.WriteLine("10. Exit");
            Console.Write("Enter your choice: ");
            choice = int.Parse(Console.ReadLine());

            // Switch statement to handle user menu choices
            switch (choice)
            {
                case 1:
                    SessionDetails();
                    break;
                case 2:
                    DisplaySession();
                    break;
                case 3:
                    IntroduceBill();
                    break;
                case 4:
                    PassBill();
                    break;
                case 5:
                    DisplayBills();
                    break;
                case 6:
                    AddOfficial();
                    break;
                case 7:
                    ViewOfficials();
                    break;
                case 8:
                    AddMLA();
                    break;
                case 9:
                    ViewMLAs();
                    break;
                case 10:
                    Console.WriteLine("Exiting...");
                    break;
                default:
                    Console.WriteLine("Invalid choice. Please try again.");
                    break;
            }
        } while (choice != 10);  // Loop until the user chooses to exit
    }

    // Method to input session details and add them to the list
    private static void SessionDetails()
    {
        Console.Write("Enter the Date of the session: ");
        string date = Console.ReadLine();
        Console.Write("Enter the Agenda of the session: ");
        string agenda = Console.ReadLine();
        sessions.Add(new Session(date, agenda));  // Add session to the list
        Console.WriteLine("Session may proceed.");
    }

    // Method to display details of all sessions
    private static void DisplaySession()
    {
        if (sessions.Count == 0)
        {
            Console.WriteLine("No session conducted yet.");
            return;
        }

        foreach (var session in sessions)
        {
            Console.WriteLine(session);
            Console.WriteLine();
        }

        // Display bills introduced in the session
        if (bills.Count == 0)
        {
            Console.WriteLine("No bills introduced in the session.");
        }
        else
        {
            Console.WriteLine("Bills introduced in the session:");
            foreach (var bill in bills)
            {
                Console.WriteLine(bill);
                Console.WriteLine();
            }
        }

        // Display officials present in the session
        if (officials.Count == 0)
        {
            Console.WriteLine("No officials to display.");
        }
        else
        {
            Console.WriteLine("Officials present in today's Session:");
            foreach (var official in officials)
            {
                Console.WriteLine(official);
                Console.WriteLine();
            }
        }

        // Display MLAs present in the session
        if (mlas.Count == 0)
        {
            Console.WriteLine("No MLAs to display.");
        }
        else
        {
            Console.WriteLine("List of MLAs present for today's session:");
            foreach (var mla in mlas)
            {
                Console.WriteLine(mla);
                Console.WriteLine();
            }
        }
    }

    // Method to introduce a new bill
    private static void IntroduceBill()
    {
        Console.Write("Enter the title of the bill: ");
        string title = Console.ReadLine();
        Console.Write("Enter the description of the bill: ");
        string description = Console.ReadLine();
        bills.Add(new Bill(title, description));  // Add bill to the list
        Console.WriteLine("Bill introduced successfully.");
    }

    // Method to pass a bill by marking it as passed
    private static void PassBill()
    {
        if (bills.Count == 0)
        {
            Console.WriteLine("No bills available to pass.");
            return;
        }

        Console.Write("Enter the title of the bill to pass: ");
        string title = Console.ReadLine();
        foreach (var bill in bills)
        {
            if (bill.ToString().Contains(title))
            {
                bill.Pass();  // Mark the bill as passed
                Console.WriteLine("Bill passed successfully.");
                return;
            }
        }
        Console.WriteLine("Bill not found.");
    }

    // Method to display all bills
    private static void DisplayBills()
    {
        if (bills.Count == 0)
        {
            Console.WriteLine("No bills to display.");
            return;
        }

        foreach (var bill in bills)
        {
            Console.WriteLine(bill);
            Console.WriteLine();
        }
    }

    // Method to add a new official to the session
    private static void AddOfficial()
    {
        Console.Write("Enter the position of the official: ");
        string position = Console.ReadLine();
        Console.Write("Enter the name of the official: ");
        string name = Console.ReadLine();
        officials.Add(new Official(position, name));  // Add official to the list
        Console.WriteLine("Official added successfully.");
    }

    // Method to display all officials
    private static void ViewOfficials()
    {
        if (officials.Count == 0)
        {
            Console.WriteLine("No officials to display.");
            return;
        }

        Console.WriteLine("List of Officials:");
        foreach (var official in officials)
        {
            Console.WriteLine(official);
            Console.WriteLine();
        }
    }

    // Method to add a new MLA to the session
    private static void AddMLA()
    {
        Console.Write("Enter the name of the MLA: ");
        string name = Console.ReadLine();
        Console.Write("Enter the constituency of the MLA: ");
        string constituency = Console.ReadLine();
        Console.Write("Enter the Political party of the MLA: ");
        string party = Console.ReadLine();
        mlas.Add(new MLA(name, constituency, party));  // Add MLA to the list
        Console.WriteLine("MLA added successfully.");
    }

    // Method to display all MLAs
    private static void ViewMLAs()
    {
        if (mlas.Count == 0)
        {
            Console.WriteLine("No MLAs to display.");
            return;
        }

        Console.WriteLine("List of MLAs:");
        foreach (var mla in mlas)
        {
            Console.WriteLine(mla);
            Console.WriteLine();
        }
    }
}

// Custom scanner class for input (not really needed as Console.ReadLine() already exists)
class Scanner
{
    public static string NextLine()
    {
        return Console.ReadLine();
    }
}
